### PR TITLE
pkg/operator: correctly sync status for the CVO

### LINKED ISSUE #346: MCO Cluster Operator statuses frequently changing
Unsure if this is causing any performance issues (it has not from what I can tell but)

The Cluster Operator machine-config-operator seems to cycle through statuses when the cluster is relatively idle (ie I am not performing and updates/applying mcs/etc...).  Not sure why the status is changing but it makes for a confusing user experience.  See the following.

`$ oc get clusteroperator -n machine-config-operator` yields varying statuses within very short periods of time.  For example:

`machine-config-operator                 3.11.0-494-g18b8dc50-dirty   True        False         False     7s`
`machine-config-operator                 3.11.0-494-g18b8dc50-dirty   False       True          False     1s`

In another example, I see changes in the `conditions:` section but the bottom entry also never changes? So the top two alternate between Available, False / Progressing, True but the oldest entry also never seems to change?

`$ oc get clusteroperator machine-config-operator -o yaml -w`
```
apiVersion: config.openshift.io/v1
kind: ClusterOperator
metadata:
  creationTimestamp: 2019-01-23T18:14:33Z
  generation: 1
  name: machine-config-operator
  resourceVersion: "919947"
  selfLink: /apis/config.openshift.io/v1/clusteroperators/machine-config-operator
  uid: bbc5a98d-1f3a-11e9-a4a4-02204cc1daaa
spec: {}
status:
  conditions:
  - lastTransitionTime: 2019-01-24T18:39:17Z
    status: "False"
    type: Available
  - lastTransitionTime: 2019-01-24T18:39:17Z
    message: Progressing towards 3.11.0-494-g18b8dc50-dirty
    status: "True"
    type: Progressing
  - lastTransitionTime: 2019-01-24T02:24:05Z
    status: "False"
    type: Failing
  extension:
    master: all 3 nodes are at latest configuration master-8b907698492f2829b35f6f826511dfa6
    worker: all 3 nodes are at latest configuration worker-58b30499568a7a2588b05ed01bba91a7
  version: 3.11.0-494-g18b8dc50-dirty
apiVersion: config.openshift.io/v1
kind: ClusterOperator
metadata:
  creationTimestamp: 2019-01-23T18:14:33Z
  generation: 1
  name: machine-config-operator
  resourceVersion: "919948"
  selfLink: /apis/config.openshift.io/v1/clusteroperators/machine-config-operator
  uid: bbc5a98d-1f3a-11e9-a4a4-02204cc1daaa
spec: {}
status:
  conditions:
  - lastTransitionTime: 2019-01-24T18:39:22Z
    status: "True"
    type: Available
  - lastTransitionTime: 2019-01-24T18:39:22Z
    status: "False"
    type: Progressing
  - lastTransitionTime: 2019-01-24T02:24:05Z
    status: "False"
    type: Failing
  extension:
    master: all 3 nodes are at latest configuration master-8b907698492f2829b35f6f826511dfa6
    worker: all 3 nodes are at latest configuration worker-58b30499568a7a2588b05ed01bba91a7
  version: 3.11.0-494-g18b8dc50-dirty
```
I'm also not seeing anything happening in the `oc logs -f` for the machine-config-operator at the times indicated in the statuses.

Does anyone have an idea why this is happening?  Why do the top two alternate, when from what I can tell nothing is happening in the MCO `oc logs -f`?

### PR DESCRIPTION
**Pull request by runcom**
Based on:

- https://github.com/openshift/cluster-version-operator/blob/master/docs/dev/clusteroperator.md#conditions
- Slack conversation with Clayton

We should correctly report Progressing only when really progressing towards something. Re-syncing, reconciling, drift detections verifying everything is still where you left it is not _changing_ something, thus is not progressing.

closes #346

/cc @smarterclayton @abhinavdahiya ptal

Signed-off-by: Antonio Murdaca <runcom@linux.com>

### CODE REVIEW

**Comment by runcom** [comment]
not entirely sure how to test this out on a running cluster, disabling the CVO isn't gonna work, setting the operator Deployment to unmanaged isn't gonna work, any idea?

**Comment by cgwalters** [comment]
If you want to test things like this cleanly it basically requires building a custom payload and passing it to the installer.

**Comment by runcom** [comment]
> If you want to test things like this cleanly it basically requires building a custom payload and passing it to the installer.

it's probably time I start doing this

File: pkg/operator/status.go [hunk 1]
```diff
@@ -31,12 +31,8 @@ func (optr *Operator) syncAvailableStatus() error {
 	// set available
 	cov1helpers.SetStatusCondition(&co.Status.Conditions, configv1.ClusterOperatorStatusCondition{
 		Type: configv1.OperatorAvailable, Status: configv1.ConditionTrue,
-		Message: fmt.Sprintf("Cluster is available at %s", optrVersion),
+		Message: fmt.Sprintf("Cluster has deployed %s", optrVersion),
 	})
-	// clear progressing
```
Comment (abhinavdahiya) [inline pkg/operator/status.go]: Why this drop?

Comment (runcom) [reply pkg/operator/status.go]: setting Available doesn't necessarily clear progressing (taken from docs https://github.com/openshift/installer/blob/master/docs/user/overview.md)
```
Failing is true with a detailed message Unable to apply 4.0.1: could not update 0000_70_network_deployment.yaml because the resource type NetworkConfig has not been installed on the server.
Available is true with message Cluster has deployed 4.0.0
Progressing is true with message Unable to apply 4.0.1: a required object is missing
```

It also looks like something other operators are doing (setting Avaiable can still be true while Failing and Progressing are also true)

Comment (runcom) [reply pkg/operator/status.go]: The other scenarios I've seen in other operators set the Conditions indipendently from the other conditions as well (which makes sense I believe)

Comment (abhinavdahiya) [reply pkg/operator/status.go]: > setting Available doesn't necessarily clear progressing (taken from docs https://github.com/openshift/installer/blob/master/docs/user/overview.md)
> 
> ```
> Failing is true with a detailed message Unable to apply 4.0.1: could not update 0000_70_network_deployment.yaml because the resource type NetworkConfig has not been installed on the server.
> Available is true with message Cluster has deployed 4.0.0
> Progressing is true with message Unable to apply 4.0.1: a required object is missing
> ```
> 
> It also looks like something other operators are doing (setting Avaiable can still be true while Failing and Progressing are also true)

you are quoting an example when operator is progressing or failing...
when the operator is available
```
```

Comment (runcom) [reply pkg/operator/status.go]: I should check the other Conditions I guess right

Comment (runcom) [reply pkg/operator/status.go]: But still, why there's an example of all three to True then? Thats really confusing.

Comment (runcom) [reply pkg/operator/status.go]: updated

Comment (runcom) [reply pkg/operator/status.go]: anyway, the status you quoted can be already converge to without my latest additions (though they're indeed needed). Progressing and Failing are set _before_ available anyway so when we're at Avaiable, we know where we are (I think)

File: pkg/operator/sync.go [hunk 2]
```diff
@@ -47,15 +47,15 @@ func (optr *Operator) syncAll(rconfig renderConfig) error {
...
 	}
 
 	agg := utilerrors.NewAggregate(errs)
 	if agg != nil {
 		errs = append(errs, optr.syncFailingStatus(agg))
 		agg = utilerrors.NewAggregate(errs)
 		return fmt.Errorf("error syncing: %v", agg.Error())
-	} else if optr.inClusterBringup {
+	}
+	if optr.inClusterBringup {
```
Comment (cgwalters) [inline pkg/operator/sync.go]: Why this change?  The intention was to stay in "initialization" stage until we'd done a successful sync.  

(Not saying it's wrong, I just want to understand why you're making the change)

File: pkg/operator/status.go [hunk 3]
```diff
@@ -28,15 +28,25 @@ func (optr *Operator) syncAvailableStatus() error {
 	}
 
 	optrVersion, _ := optr.vStore.Get("operator")
+
+	var progressing bool
+	if !optr.vStore.Equal(co.Status.Versions) {
+		progressing = true
```
Comment (cgwalters) [inline pkg/operator/status.go]: One thing I'm wondering here...we want `progressing=true` when `optr.inClusterBringup` too right?
IOW this code is handling the upgrade case but we also want to be progressing during initial install.

File: pkg/operator/sync.go (hunk 2 above)
Comment (runcom) [reply pkg/operator/sync.go]: just golang linting, the previous if branch has a return at the end, the `else` branch is therefore superfluous

File: pkg/operator/status.go (hunk 3 above)
Comment (runcom) [reply pkg/operator/status.go]: good call, I'm gonna work that out

Comment (runcom) [reply pkg/operator/status.go]: fixed here https://github.com/openshift/machine-config-operator/pull/406/files#diff-bee1f8f36240cb95db10244fdf335146R68 - we now set progressing = true on bring up as well

File: pkg/operator/status.go (hunk 1 above)
Comment (runcom) [reply pkg/operator/status.go]: fixed here I believe now https://github.com/openshift/machine-config-operator/pull/406/files#diff-bee1f8f36240cb95db10244fdf335146R37

**Comment by runcom** [comment]
This is good to review and test now, I was able to test it out using a custom payload built using https://github.com/openshift/machine-config-operator/issues/421#issuecomment-463392262 and I can't notice the statuses bouncing anymore.

**Comment by runcom** [comment]
@abhinavdahiya could you also take another look at this now?

File: pkg/operator/status.go (hunk 3 above)
Comment (cgwalters) [reply pkg/operator/status.go]: One thing I've noticed related to this though is that we seem to stay in `inClusterBringup` for multiple times.  Though maybe that's only failing PRs?  Glancing at [the logs in a recent merged PR https://gcsweb-ci.svc.ci.openshift.org/gcs/origin-ci-test/pr-logs/pull/openshift_machine-config-operator/419/pull-ci-openshift-machine-config-operator-master-e2e-aws/1756/artifacts/e2e-aws/pods/] things look fine.

File: pkg/operator/status.go [hunk 4]
```diff
@@ -28,15 +28,21 @@ func (optr *Operator) syncAvailableStatus() error {
 	}
 
 	optrVersion, _ := optr.vStore.Get("operator")
+	progressing := cov1helpers.IsStatusConditionTrue(co.Status.Conditions, configv1.OperatorProgressing)
+	failing := cov1helpers.IsStatusConditionTrue(co.Status.Conditions, configv1.OperatorFailing)
+	message := fmt.Sprintf("Cluster has deployed %s", optrVersion)
+	status := configv1.ConditionTrue
```
Comment (cgwalters) [inline pkg/operator/status.go]: Let's name this variable `available`

File: pkg/operator/status.go [hunk 5]
```diff
@@ -55,15 +61,20 @@ func (optr *Operator) syncProgressingStatus() error {
 	}
 
 	optrVersion, _ := optr.vStore.Get("operator")
-	var message string
+	status := configv1.ConditionFalse
```
Comment (cgwalters) [inline pkg/operator/status.go]: Ditto.

File: pkg/operator/status.go [hunk 6]
```diff
@@ -55,15 +61,20 @@ func (optr *Operator) syncProgressingStatus() error {
...
+	message := fmt.Sprintf("Cluster version is %s", optrVersion)
+
 	if optr.vStore.Equal(co.Status.Versions) {
-		// syncing the state to existing version.
-		message = fmt.Sprintf("Running resync for %s", optrVersion)
+		if optr.inClusterBringup {
+			status = configv1.ConditionTrue
+		}
```
Comment (cgwalters) [inline pkg/operator/status.go]: Wait why would we be `available = false` if the version store is equal and `!inClusterBringup`?

File: pkg/operator/status.go (hunk 3 above)
Comment (runcom) [reply pkg/operator/status.go]: mmm what do you mean "stay in `inClusterBringup` for multiple times"?

Comment (cgwalters) [reply pkg/operator/status.go]: If the sync fails; you'll see that log multiple times.  But I think that's the right thing?

File: pkg/operator/status.go (hunk 6 above)
Comment (runcom) [reply pkg/operator/status.go]: this is setting Progressing to false, not available. If `!inClusterBringup && version-equal`, `available is gonna be true (assuming failing is false as well). Am I missing something?

Comment (cgwalters) [reply pkg/operator/status.go]: Sorry, yeah.  I guess it's clearer when viewing the surrounding code as the functions clearly say `syncProgressingStatus` and `syncAvailableStatus`.

So nevermind, LGTM!

File: pkg/operator/status.go (hunk 5 above)
Comment (runcom) [reply pkg/operator/status.go]: fixed

File: pkg/operator/status.go (hunk 4 above)
Comment (runcom) [reply pkg/operator/status.go]: fixed

File: pkg/operator/status.go (hunk 3 above)
Comment (runcom) [reply pkg/operator/status.go]: oh yeah, it's correct we keep re-syncing and logging that

**Comment by abhinavdahiya** [comment]
@runcom @cgwalters 
With this PR if MCO ever fails it will never go available again because of
https://github.com/openshift/machine-config-operator/pull/406/files#diff-bee1f8f36240cb95db10244fdf335146R37

example:
https://gubernator.k8s.io/build/origin-ci-test/logs/release-openshift-origin-installer-e2e-aws-4.0/4246
MCO is reporting failing...
but if you look at https://storage.cloud.google.com/origin-ci-test/logs/release-openshift-origin-installer-e2e-aws-4.0/4246/artifacts/release-e2e-aws/pods/openshift-machine-config-operator_machine-config-operator-7888cd6-c8wp4_machine-config-operator.log.gz

mco looses the connection to apiserver interminently when reconciling and then reports the machine-config failing.

once MCO has reported failing, even on a successful reconcile later on calling syncAvailbleStatus [here](https://github.com/openshift/machine-config-operator/pull/406/files#diff-fa45321336db7ad1cedc28bf643a4f97R63) will still report failing because of that `if` condition I mentioned above.

Please look into this.

**Comment by abhinavdahiya** [comment]
I also suspect this was similar error https://gubernator.k8s.io/build/origin-ci-test/logs/release-openshift-origin-installer-e2e-aws-4.0/4197

**Comment by runcom** [comment]
Thanks @abhinavdahiya I'm fixing that (I was also thinking about making all of this simpler as well but later).

**Comment by runcom** [comment]
@abhinavdahiya please take a look at https://github.com/openshift/machine-config-operator/pull/442
